primitives/src/crh/poseidon/mod.rs

```diff
@@ -17,109 +21,105 @@ pub mod batched_crh;
+        let mut new_state = vec![<Self as FieldBasedHashParameters>::Fr::zero(); Self::T];
 
-    let mut idx_cst = 0;
-    for i in 0..P::T {
-        scalar_mul::<F,P>(&mut new_state[i], state, idx_cst);
-        idx_cst += P::T;
+        let mut idx_cst = 0;
+        for i in 0..Self::T {
+            Self::scalar_mul(&mut new_state[i], state, idx_cst);
+            idx_cst += Self::T;
+        }
+        *state = new_state;
     }
-    *state = new_state;
 }
 
-// Function that does the scalar multiplication
-// It uses a partial Montgomery multiplication defined as PM(x, t) = x * t * 2^-64 mod M
-// t is a 64-bit matrix constant. In the algorithm, the constants are represented in
-// partial Montgomery representation, i.e. t * 2^64 mod M
-#[inline]
-pub fn scalar_mul_fast<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr=F>> (res: &mut F, state: &mut[F], mut start_idx_cst: usize) {
-    state.iter().for_each(|&x| {
-        let elem = P::MDS_CST_SHORT[start_idx_cst].mul_short(x);
-        start_idx_cst += 1;
-        *res += &elem;
-    });
+pub trait PoseidonShortParameters: PoseidonParameters {
+    /// MDS matrix supporting short Montgomery multiplication with respect to the short
+    /// Montgomery constant R_2=2^64
+    const MDS_CST_SHORT: &'static [Self::Fr];
 }
 
-// Function that does the mix matrix with fast algorithm
-#[inline]
-pub fn matrix_mix_short<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr=F>> (state: &mut Vec<F>) {
-
-    // the new state where the result will be stored initialized to zero elements
-    let mut new_state = vec![F::zero(); P::T];
-
-    let mut idx_cst = 0;
-    for i in 0..P::T {
-        scalar_mul_fast::<F,P>(&mut new_state[i], state, idx_cst);
-        idx_cst += P::T;
-    }
-    *state = new_state;
+#[derive(Derivative)]
+#[derivative(
+Clone(bound = ""),
+Debug(bound = ""),
+)]
+pub struct PoseidonHash<F: PrimeField, P: PoseidonParameters<Fr = F>, SB: SBox<Field = F, Parameters = P>>{
+    state: Vec<F>,
+    pending: Vec<F>,
+    input_size: Option<usize>,
+    updates_ctr: usize,
+    mod_rate: bool,
+    _parameters: PhantomData<P>,
+    _sbox: PhantomData<SB>,
 }
 
-impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr=F>> PoseidonHash<F, P> {
-
+impl<F, P, SB> PoseidonHash<F, P, SB>
+    where
+        F: PrimeField,
+        P: PoseidonParameters<Fr = F>,
+        SB: SBox<Field = F, Parameters = P>,
+{
     fn _init(constant_size: Option<usize>, mod_rate: bool, personalization: Option<&[F]>) -> Self
     {
         let mut state = Vec::with_capacity(P::T);
         for i in 0..P::T {
             state.push(P::AFTER_ZERO_PERM[i]);
         }
 
-
         let mut instance = Self {
             state,
             pending: Vec::with_capacity(P::R),
             input_size: constant_size,
             updates_ctr: 0,
             mod_rate,
             _parameters: PhantomData,
+            _sbox: PhantomData,
```

Comment:
I addressed that in one of my comments on PR #69 : The personalization padding few lines below allows trivial collisions. We need to change to the 1,0,...0 padding as from the Poseidon paper.

Reply:
Will be changes in PR #69.